### PR TITLE
[GPT-3]Add pure fp16 for dygraph

### PR DESCRIPTION
**Pull request by ForFishes**
### PR types

 New features

### PR changes

Others
### Description

[GPT-3]Add pure fp16 for dygraph

### CODE REVIEW

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 1]
```diff
@@ -192,15 +193,21 @@ def do_train(args):
...
 
-    if paddle.distributed.get_world_size() > 1:
-        model = fleet.distributed_model(model)
-        optimizer = fleet.distributed_optimizer(optimizer)
 
-    if args.use_amp:
+    if args.use_pure_fp16:
         scaler = paddle.amp.GradScaler(init_loss_scaling=args.scale_loss)
         scaler = fleet.distributed_scaler(scaler)
+        model, optimizer = paddle.amp.decorator(models=model,
```
Comment (zhangbo9674) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:203]: now api is paddle.amp.decorate

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 2]
```diff
@@ -192,15 +193,21 @@ def do_train(args):
...
-        model = fleet.distributed_model(model)
-        optimizer = fleet.distributed_optimizer(optimizer)
 
-    if args.use_amp:
+    if args.use_pure_fp16:
         scaler = paddle.amp.GradScaler(init_loss_scaling=args.scale_loss)
         scaler = fleet.distributed_scaler(scaler)
+        model, optimizer = paddle.amp.decorator(models=model, 
+                                                optimizers=optimizer, 
+                                                mode='L2',
```
Comment (zhangbo9674) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:205]: level='O2'

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 3]
```diff
@@ -245,20 +252,16 @@ def do_train(args):
...
-                            custom_white_list=[
-                                "layer_norm", "softmax", "gelu"
-                            ],
+                            args.use_pure_fp16,
                             custom_black_list=[
-                                "reduce_sum", "c_softmax_with_cross_entropy",
-                                "c_embedding"
-                            ]):
+                                "reduce_sum", "c_softmax_with_cross_entropy","elementwise_div"
+                            ], mode='L2'):
```
Comment (zhangbo9674) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:258]: level='O2'

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 4]
```diff
@@ -270,19 +273,15 @@ def do_train(args):
...
```
Comment (zhangbo9674) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:279]: level='02'

**Comment by ZeyuChen** [comment]
@zhangbo9674 Is this PR still work in progress? @ForFishes

**Comment by ZHUI** [comment]
The pure fp 16 have been supported by PR https://github.com/PaddlePaddle/PaddleNLP/pull/1176 ,  so we close this PR for now.

**Pull request closed without merging**